ecosystem/promise.md

```diff
@@ -0,0 +1,141 @@
+    mapping(bytes32 => Callback[]) callbacks;
+
+    /// @notice attach a continuation dependent only on the return value of the remote message
+    function then(bytes32 _msgHash, bytes4 _selector) external {
+        callbacks[_msgHash].push(Callback({ target: msg.sender, selector: _selector, context: ""}))
+        emit CallbackRegistered(_msgHash);
+    }
+
+    /// @notice attach a continuation dependent on the return value and some additional saved context
+    function then(bytes32 _msgHash, bytes4 _selector, bytes memory _context) external {
+        callbacks[_msgHash].push(Callback({ target: msg.sender, selector: _selector, context: _context}))
+        emit CallbackRegistered(_msgHash);
+    }
+}
+```
+
+The updated `RelayedMessage` event can be used to invoke these continuations by a relayer.
+
+```solidity
+library Promise {
+
+    /// @notice invoke continuations present on the completion of a remote message
+    function continue(Identifier calldata _id, bytes calldata payload) external payable {
+        if (_id.origin != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert InvalidOrigin();
+        ICrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(_id, keccak256(_payload));
+
+        bytes32 eventSel = abi.decode(_payload[:32], (bytes32));
+        if (eventSel != IL2ToL2CrossDomainMessenger.RelayedMessage.selector) revert InvalidEvent();
+
+        (,, bytes32 msgHash, bytes memory returnData) = abi.decode(_payload[32:], (uint256, uint256, bytes32, bytes));
+
+        for (uint256 i = 0; i < callbacks[msgHash].length; i++) {
+            Callback memory callback = callbacks[msgHash];
+            bytes memory data = callback.context.length > 0 ?
+                abi.encode(callback.selector, returnData, callback.context) :
+                abi.encode(callback.selector, returnData);
+
+            (completed,) = target.call(data);
+            require(completed)
+        }
+
+        delete callbacks[msgHash];
+        emit CallbacksCompleted(msgHash);
+    }
+}
+```
+
+![Promise Flow Diagram](./promise/promise.png)
+
+## Notes
+
+- Requires no fundamental contract API changes to the CDM besides the event.
+- Different kinds of callbacks can be introduced on the source, all without change to the destination. `onFailure`, `onTimeout`, etc. The log identifier for the event has timestamp information. Currently the `L2ToL2CrossDomainMessenger` does not allow for failed message delivery but if it did, failure callbacks could be created as well.
+- Multiple callbacks can be registered to the same message. `then()` and `onTimeout()`.
+- Relayers must listen to `CallbackRegistered` events to know when which messages have pending continuations. The tradeoff being additional touch-points for relayers to hook into. In the alternative, if a message was to be sent back from the destination, it would be almost be a no-op in relayer infra since the message outbox would look the same.
+
+# Alternatives Considered
+
+## Callback Entrypoint
```

Comment:
The entrypoint solution and this proposal are not mutually exclusive though, right? You are simply proposing that the CDM emit return data that can be used by other contracts if I understand correctly. Entrypoints would still have a place but wouldn't be the preferred way to address this use case?